reconcile/utils/gql.py

```diff
@@ -229,29 +263,61 @@ def init(
     commit: Optional[str] = None,
     commit_timestamp: Optional[str] = None,
 ):
-    global _gqlapi
-    _gqlapi = GqlApi(
+    return GqlApiSingleton(
         url,
         token,
         integration,
         validate_schemas,
         commit=commit,
         commit_timestamp=commit_timestamp,
-    )
-    return _gqlapi
+    ).gql_api
 
 
 def get_resource(path: str) -> dict[str, Any]:
     return get_api().get_resource(path)
 
 
-def _init_gql_client(url: str, token: Optional[str]) -> Client:
-    req_headers = None
-    if token:
-        # The token stored in vault is already in the format 'Basic ...'
-        req_headers = {"Authorization": token}
-    # Here we are explicitly using sync strategy
-    return Client(transport=RequestsHTTPTransport(url, headers=req_headers, timeout=30))
+class PersistentRequestsHTTPTransport(RequestsHTTPTransport):
+    """A transport for the GQL Client that uses an existing.
+    Is a reduced version of the RequestsHTTPTransport class from gql library
+    with the connect and close methods removed, cause they are implemented
+    to disconnect after each query.
+    """
+
+    def __init__(
+        self,
+        session: requests.Session,
+        url: str,
+        headers: Optional[Dict[str, Any]] = None,
+        cookies: Optional[Union[Dict[str, Any], RequestsCookieJar]] = None,
+        auth: Optional[AuthBase] = None,
+        use_json: bool = True,
+        timeout: Optional[int] = None,
+        verify: Union[bool, str] = True,
+        retries: int = 0,
+        method: str = "POST",
+        **kwargs: Any,
+    ):
+        super().__init__(
+            url,
+            headers,
+            cookies,
+            auth,
+            use_json,
+            timeout,
+            verify,
+            retries,
+            method,
+            **kwargs,
+        )
+        # can't directly assign, due to mypy type checking
+        self._session = session
```

Comment:
if we don't assign it in `__init__`, then there will be an edge case that, when `close` is called, `self.session` is None. https://github.com/app-sre/qontract-reconcile/pull/3741/files#diff-ebb669f2f6dc69ceb880efd75f362c14513914fb8737d013dd98f82351964672R140 will break.

Reply:
Close just has `pass`. What should close the session here?

Reply:
not that close
```
self.client.transport.session
```
will be None

Reply:
will add 
```
# type: ignore
```

Reply:
Concluded, will change the implementation again to always reset the singleton, cause some uses cases like early exist rely on recreating the API instance with changed parameters